apiserver/facades/agent/controllercharm/controllercharm.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controllercharm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/environs/bootstrap"
+	"github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/state"
+)
+
+const (
+	// jujuMetricsUserPrefix defines the "namespace" in which this facade is
+	// allowed to create/remove users.
+	jujuMetricsUserPrefix = "juju-metrics-"
+
+	// userCreator is the listed "creator" of metrics users in state.
+	userCreator = "admin"
+)
+
+// API provides API methods to the controllercharm worker.
+type API struct {
+	state backend
+}
+
+// backend defines the state methods that API needs.
+type backend interface {
```

Comment:
Not sure why we define these private interfaces and create the `fakeState` for tests instead of using gomock.

Reply:
Well, I would ask the opposite question. Why use `gomock` and deal with the extra overhead of mocks, generated/opaque code we don't understand, when we can easily write a simple stub to test the functionality here.

As these interfaces are not used outside this package, they should be private.